rfcs/0145-doc-strings.md

```diff
@@ -0,0 +1,500 @@
+- Standardize a format for doc comments that further RFCs can extend
+  - E.g. extend with code examples that can be converted to tests (aka doctests).
+
+This RFC is a significant change to existing documentation conventions.
+It allows distinguishing between regular and doc comments. Having distinction is essential because arbitrary code comments should not end up in generated documentation.
+
+> Hint: Generating static documentation is controvert topic in nixpkgs. We found that it is impossible to generate accurate documentation statically. A correct solution would involve the evaluation of expressions in some way.
+
+## Current State
+
+Currently, Nix does not have doc comments. Over time there evolved certain conventions, but they were never formally specified.
+Within this RFC, we want to clarify which convention we should use, harmonize and improve the format for doc-comments to achieve the envisioned goals.
+
+We use a tool called [nixdoc](https://github.com/nix-community/nixdoc) to use a 'doc-comment'- like functionality, which allows us to build a subset of static documentation of the `nixpkgs.lib` functions.
+
+The format of a doc-comment was, however, never specified. The nix community does not understand the rules enforced by `nixdoc`, which we believe is one of the reasons for its lack of adoption.
+
+Here is an example of the format understood by *nixdoc*:
+
+```nix
+# nixpkgs/lib/attrsets.nix
+
+{
+  /* <Description Field>
+     Example:
+     <Comprehensive Code example>
+     Type:
+     <Type signature>
+  */
+  AttrFunc =
+    # <Desribe arg1>
+    arg1:
+    # <Describe arg2>
+    arg2:
+    
+    # ... implementation
+}
+```
+
+## Current problems
+
+### Lack of having an agreed format
+
+The Nix community interprets the outer format and content of doc-comments differently. Within nixpkgs, several conventions and patterns still need to be harmonized. Some are compatible with nixdoc, while others do not and decide upon their conventions. We immediately observe one clear thing: Most of them utilize some form of markdown, even if the content is not rendered.
+
+In general, the format for writing documentation strings is **not specified**.
+
+The *nixdoc*-tool requires a somewhat consistent format, but the basic format was never specified and cannot be enforced in parts of nixpkgs where nixdoc is currently not applied.
+
+Extending the scope of *nixdoc* is not the primary goal. Instead, we should find formal rules for writing *doc-comments*. Tools like *nixdoc* can then implement against this RFC instead of the format relying on nixdoc implementation details.
+
+### Impossible to differentiate from regular comments
+
+The format does not allow any distinction between doc comments and regular comments.
+
+Having a distinction would allow us to
+
+1. Find all comments that are part of the documentation
+2. Render them in the documentation format
+3. Connect the documentation to the exact places in the Nix code. 
```

Comment:
These are all possible without introducing a distinct format for doc-comments though, right?

Reply:
I just wanted to let you know that I can follow your arguments.

> Couldn't this RFC be simplified by just stating that doc-comments:
>   - are block comments immediately before an attribute
>   - are formatted with CommonMark
>  - have Type and Example fields

# Are block comments immediately before an attribute?

How would you find all doc comments if you cannot distinguish them from regular comments?
So this would mean all multiline comments are doc comments if they follow specific placement rules.
I wanted to avoid that and follow Rust's approach by adding one more character, clearly showing "this is a doc-comment". Otherwise, you would make multiline comments almost impossible for any "non-doc-purpose". 

Did you want to suggest that every multiline comment is always a doc-comment, and there are no regular multiline comments anymore?

# Are formatted with commonMark?

Finally, that is a point that all agreed on. :)

# Have Type and Example fields?

I recently joined one of the documentation team meetings, and we discussed that a bit.
The outcome was that (1) example and type should also be commonMark headings.
(2) this RFC should not specify those fields in advance. It might be a convention in nixpkgs. If we'd only allow `Example` and `Type`, that would be pretty restrictive, and I've seen some PRs and threads where people used other headings. So the best choice we'd come up with is to say: "the entire content is commonMark". This is the least restrictive and still allows us to follow that convention without enforcing it.

Reply:
About your first questions: how does Go find all doc-comments? AFAIU, they just assume that all comments before a to-level literal are doc-comments, and the `go doc` tool extacts them accordingly. From the docs:

> “Doc comments” are comments that appear immediately before top-level package, const, func, type, and var declarations with no intervening newlines. Every exported (capitalized) name should have a doc comment.

> The [go doc](https://go.dev/cmd/go#hdr-Show_documentation_for_package_or_symbol) command looks up and prints the doc comment for a given package or symbol. (A symbol is a top-level const, func, type, or var.)

About your third: what other headings have you seen in the wild? I remember we tried to come up with something other than `Example` and `Type` and we couldn't. Would be curious to see what you've found.

Reply:
> About your third: what other headings have you seen in the wild? I remember we tried to come up with something other than `Example` and `Type` and we couldn't. Would be curious to see what you've found.

Given that `rustdoc` has been referenced, it might be interesting to compare Rust RFCs [505](https://github.com/rust-lang/rfcs/blob/ef26c15140f0530ad05d59253f0527e5384a7454/text/0505-api-comment-conventions.md) and [1574](https://github.com/rust-lang/rfcs/blob/ef26c15140f0530ad05d59253f0527e5384a7454/text/1574-more-api-documentation-conventions.md), which, among other things, define a standard set of such headings for Rust: Examples, Panics, Errors, Safety, Aborts, Undefined Behavior.  I think the "Safety" and "Undefined Behavior" would be unneeded for Nix.  "Errors", for Nix, would say what return values of a function count as error values and under what conditions the function will return those values.  "Panics", for Nix, would be "Exceptions" and say what exceptions a function may `throw` and under what conditions it will `throw` them.  "Aborts" would be the same but for `abort`.  It may make sense to merge "Exceptions" and "Aborts".  Assertions imply a type of abort.  Functions that don't use `throw`, `abort`, or `assert` would omit these sections.

Reply:
I think this RFC doesn't specify any further format inside the doc comments (other that it's commonmark). What fields to have in there, if at all, should be another bikeshed after this PR IMHO.